sdk/include/opentelemetry/sdk/common/global_log_handler.h

```diff
@@ -137,24 +137,25 @@ OPENTELEMETRY_END_NAMESPACE
-    using opentelemetry::sdk::common::internal_log::LogHandler;                           \
-    if (level > GlobalLogHandler::GetLogLevel())                                          \
-    {                                                                                     \
-      break;                                                                              \
-    }                                                                                     \
-    const opentelemetry::nostd::shared_ptr<LogHandler> &log_handler =                     \
-        GlobalLogHandler::GetLogHandler();                                                \
-    if (!log_handler)                                                                     \
-    {                                                                                     \
-      break;                                                                              \
-    }                                                                                     \
-    std::stringstream tmp_stream;                                                         \
-    tmp_stream << message;                                                                \
-    log_handler->Handle(level, __FILE__, __LINE__, tmp_stream.str().c_str(), attributes); \
+#define OTEL_INTERNAL_LOG_DISPATCH(level, message, attributes)                      \
+  do                                                                                \
+  {                                                                                 \
+    using opentelemetry::sdk::common::internal_log::GlobalLogHandler;               \
+    using opentelemetry::sdk::common::internal_log::LogHandler;                     \
+    if (level > GlobalLogHandler::GetLogLevel())                                    \
+    {                                                                               \
+      break;                                                                        \
+    }                                                                               \
+    const opentelemetry::nostd::shared_ptr<LogHandler> &log_handler =               \
+        GlobalLogHandler::GetLogHandler();                                          \
+    if (!log_handler)                                                               \
+    {                                                                               \
+      break;                                                                        \
+    }                                                                               \
+    std::stringstream tmp_stream;                                                   \
+    tmp_stream << message;                                                          \
+    std::string tmp_string = tmp_stream.str();                                      \
+    log_handler->Handle(level, __FILE__, __LINE__, tmp_string.c_str(), attributes); \
```

Comment:
Probably there is something I missed. What's is different behavior with this change?

Reply:
I also didn't get the point, could you please explain this change?
To my understanding, C++ specifications will ensure the temporary object should be freed only after this code block(`Handle(...)`).

> https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime

Reply:
As you mentioned, according to 
https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime,
"All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created." 
Therefore, after the execution of the argument `tmp_stream.str().c_str()`, the temporary object `tmp_stream.str()` will be destructed, and the parameter that log_handler->Handle receives will be an invalid pointer. 
When a function returns a string, the c_str method of the returned temporary string object cannot be invoked. This issue is discussed here: 
https://stackoverflow.com/questions/27627413/why-does-calling-c-str-on-a-function-that-returns-a-string-not-work 
So, what I'm pointing out here is not that log_handler is using the string asynchronously, but that even in synchronous situations, tmp_stream.str().c_str() is problematic.

Reply:
You're correct. 

We've previously encountered a core dump caused by c_str(), similar to the situation described in https://stackoverflow.com/questions/27627413/why-does-calling-c-str-on-a-function-that-returns-a-string-not-work.
The code was as follows:
```
    auto cstr = tmp_stream.str().c_str();
    f(cstr);
```
This caused `cstr` to become an invalid pointer passed into function f.

However, in this case, Handle(level, __FILE__, __LINE__, tmp_stream.str().c_str(), attributes) is within a single expression, so this issue doesn't exist.

When I saw this core dump, my habitual aversion to the c_str() function led to a misjudgment.

Reply:
Yes, in the example you shared and aslo the stackoverflow link, the temporary object returned from function reaches the end of its lifetime at the end of the including its expression, which means the temporary object is no longer valid after the ";", but the lifetime of local variable is different and will reach the end of current block, like @owent mentioend.